src/main/scala/edu/cmu/cs/obsidian/parser/AST.scala

```diff
@@ -64,95 +89,182 @@ sealed abstract class InvokableDeclaration() extends Declaration {
+    override val obstype: Option[ObsidianType] = e.obstype // todo
+
     override def toString: String = {
         s"$e.$f"
     }
 }
 
 case class LocalInvocation(name: String, genericParams: Seq[GenericType],
-                           params: Seq[ObsidianType], args: Seq[Expression]) extends Expression {
+                           params: Seq[ObsidianType], args: Seq[Expression], typ: Option[ObsidianType]) extends Expression {
+    override val obstype: Option[ObsidianType] = typ
+
     override def substitute(genericParams: Seq[GenericType], actualParams: Seq[ObsidianType]): LocalInvocation =
         LocalInvocation(name,
             genericParams,
             params.map(_.substitute(genericParams, actualParams)),
-            args.map(_.substitute(genericParams, actualParams)))
+            args.map(_.substitute(genericParams, actualParams)), typ)
             .setLoc(this)
 
     override def toString: String = s"$name(${args.mkString(",")})"
 }
+
 case class Invocation(recipient: Expression, genericParams: Seq[GenericType], params: Seq[ObsidianType],
-                      name: String, args: Seq[Expression], isFFIInvocation: Boolean) extends Expression {
+                      name: String, args: Seq[Expression], isFFIInvocation: Boolean, typ: Option[ObsidianType]) extends Expression {
+    override val obstype: Option[ObsidianType] = typ
+
     override def toString: String = s"$recipient.$name(${args.mkString(",")})"
 
     override def substitute(genericParams: Seq[GenericType], actualParams: Seq[ObsidianType]): Invocation =
         Invocation(recipient.substitute(genericParams, actualParams),
             genericParams,
             params.map(_.substitute(genericParams, actualParams)),
             name,
-            args.map(_.substitute(genericParams, actualParams)), isFFIInvocation)
+            args.map(_.substitute(genericParams, actualParams)), isFFIInvocation, typ)
             .setLoc(this)
 }
-case class Construction(contractType: ContractType, args: Seq[Expression], isFFIInvocation: Boolean) extends Expression {
+
+case class Construction(contractType: ContractType, args: Seq[Expression], isFFIInvocation: Boolean, typ: Option[ObsidianType]) extends Expression {
+    override val obstype: Option[ObsidianType] = typ
+
     override def substitute(genericParams: Seq[GenericType], actualParams: Seq[ObsidianType]): Construction =
         Construction(contractType.substitute(genericParams, actualParams),
-            args.map(_.substitute(genericParams, actualParams)), isFFIInvocation)
+            args.map(_.substitute(genericParams, actualParams)), isFFIInvocation, typ)
             .setLoc(this)
 }
-case class Disown(e: Expression) extends UnaryExpression(Disown, e)
-case class StateInitializer(stateName: Identifier, fieldName: Identifier) extends Expression {
+
+case class Disown(e: Expression) extends UnaryExpression(Disown, e) // todo there's something happening that i don't understand
```

Comment:
Namely?

Reply:
I spent some time trying to add a `obstype` field to this case class, overriding the `None` I think it would inherit from `UnaryExpression` with the obstype of `e`, but I was struggling to get the syntax right to express what I meant. `UnaryExpression` is parametric (that's probably not the scala-specific word) in first parameter, which is then always recursive here, so I don't quite know how to change it --- or if I should.